src/utils/parse.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{model::prelude::*, prelude::*};
+
+// Questions:
+// - Should the error enums be #[non_exhaustive]?
+// - Should the MessageParseError implement source()? It may be too much boilerplate (maybe add thiserror dependency?)
+
+/// Parse a value from a string in context of a received message.
+///
+/// This trait is a superset of [`std::str::FromStr`] (the trait used in [`str::parse`]). The
+/// difference is that this trait supports serenity-specific Discord types like [`Member`],
+/// [`Role`] (soon), or [`Emoji`] (soon).
+///
+/// Trait implementations may do network requests as part of their parsing procedure.
+///
+/// Useful for implementing command argument parsing frameworks.
+#[async_trait::async_trait]
+pub trait Parse: Sized {
+    /// The associated error which can be returned from parsing.
+    type Err;
+
+    /// Parses a string `s` as a command parameter of this type.
+    async fn parse(ctx: &Context, msg: &Message, s: &str) -> Result<Self, Self::Err>;
+}
+
+#[async_trait::async_trait]
+impl<T: std::str::FromStr> Parse for T {
+    type Err = <T as std::str::FromStr>::Err;
+
+    async fn parse(_: &Context, _: &Message, s: &str) -> Result<Self, Self::Err> {
+        T::from_str(s)
+    }
+}
+
+/// Error that can be returned from [`Member::parse`].
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum MemberParseError {
+    GuildNotInCache,
+    NotFoundOrMalformed,
+}
+
+impl std::error::Error for MemberParseError {}
+
+impl std::fmt::Display for MemberParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+			Self::GuildNotInCache => write!(f, "guild is not in cache"),
+			Self::NotFoundOrMalformed => write!(f, "provided member was not found or provided string did not adhere to any known guild member format"),
+		}
+    }
+}
+
+/// Look up a guild member by a string case-insensitively.
+///
+/// Requires the cache feature to be enabled.
+///
+/// The lookup strategy is as follows (in order):
+/// 1. Lookup by ID.
+/// 2. Lookup by mention.
+/// 3. Lookup by name#discrim
+/// 4. Lookup by name
+/// 5. Lookup by nickname
+#[cfg(feature = "cache")]
+#[async_trait::async_trait]
+impl Parse for Member {
+    type Err = MemberParseError;
+
+    #[doc = ""] // to get impl documentation to show
```

Comment:
Comments go before the related code item.
```suggestion
    // to get impl documentation to show
    #[doc = ""] 
```